refactor(semantic): make SemanticBuilder opaque

DonIsaac opened the pull request:
change visibility of building-related methods and properties of SemanticBuilder
from `pub` to `pub(crate)`.

DonIsaac commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4851?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4851** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4851?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4850** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4850?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

rzvxa commented:
This is lovely, It's so annoying having builder fields available from the outside world. But let's not merge it until we also get our other breaking changes ready.
Adding `AstNodeId` and reordering call expression fields are also breaking changes so it would be worth it to wait just so we can increase the minor version once for all of these.

Boshen commented:
Let me test Rolldown before merging.

overlookmotel commented:
Very much agree with this change. If Rolldown is relying on accessing properties/methods of `SemanticBuilder`, it shouldn't be, and we should provide alternative methods on `Semantic` for them to access the info they need.

DonIsaac committed:
refactor(semantic): make SemanticBuilder opaque (#4851)

change visibility of building-related methods and properties of SemanticBuilder
from `pub` to `pub(crate)`.
